Redirect new/edit/delete URLs from CPT implementation to COT

jorgeatorres opened the pull request:
### All Submissions:

-   [X] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [X] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
This PR adds a controller class that takes care of redirecting requests for certain order-related actions from their CPT version to their COT version.

Closes #34229.

### How to test the changes in this Pull Request:

1. Make posts authoritative and make sure to populate your site with some orders.
2. Go to WC > Orders and take note of the URL as well as some of the links to specific orders and/or order actions such as "trash" or "untrash".
3. Make COT authoritative and migrate orders.
4. Try to use some of the links from step 2 and confirm that you are redirected to the new URLs (i.e. those under `admin.php?page=wc-orders`). In most cases, query args are preserved verbatim, except for some specific ones that are translated to new ones or removed entirely (for example, `post_type` is no longer needed).
   Some scenarios that could be tested:
   - `/wp-admin/post-new.php?post_type=shop_order` should take you to the new order screen (/wp-admin/admin.php?page=wc-orders&action=new).
   - `/wp-admin/post.php?post=POST_ID&action=edit` should take you to the edit screen for `POST_ID`. If `POST_ID` is not an order, no redirection should take place.
   - `/wp-admin/edit.php?post_type=shop_order` should take you to the main order screen.
   - `/wp-admin/edit.php?post_status=wc-pending&post_type=shop_order` should take you to the main order screen with `status=wc-pending`.
   - `/wp-admin/edit.php?s&post_status=all&post_type=shop_order&action=-1&m=202208&_customer_user&filter_action=Filter&paged=1&action2=-1` should take you to the main order screen with all query args still present (and thus, with the date filter active to Aug 2022).
   - `wp-admin/post.php?post=POST_ID&action=trash&_wpnonce=NONCE` should trash the order using COT and take you to the orders screen. Keep in mind that `NONCE` can expire, so this kind of link should be tested quickly.

### Other information:

-   [X] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [X] Have you successfully run tests with your changes locally?
-   [X] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

jorgeatorres committed:
Add utility method to generate a link to the orders screen

jorgeatorres committed:
Add `PostsRedirectionController` to handle redirection from CPT-based URLs to new ones

jorgeatorres committed:
Use the redirections controller inside the orders screen PageController

jorgeatorres committed:
Add changelog

jorgeatorres committed:
Add check for COT enabled before enabling redirection controller

jorgeatorres committed:
Add support for trashing orders

barryhughes commented:
Thanks!

The only thing I am having trouble with are bulk actions (such as trashing orders).

I don't think the problem is that the nonce has expired, though, I think it's because the nonce actions are different (the CPT editor uses an action of `bulk-posts` and the bulk action handler in the COT implementation checks for `bulk-orders`). Possibly we also needed handling for the admin referer string?

Let me know if I'm misunderstanding/testing incorrectly.

barryhughes commented on plugins/woocommerce/changelog/fix-34229:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: enhancement
+
+Redirect CPT-based URLs to OrdersTableDataStore ones.
```
```suggestion
Redirect CPT-based admin order URLs to the corresponding Custom Order Table URLs.
```

Possible re-wording that I think might be clearer to consumers of the changelog. Optional (since, it's a bit of a subjective thing)—feel free to skip if you prefer the existing wording.

barryhughes commented on plugins/woocommerce/src/Internal/Admin/Orders/PostsRedirectionController.php:
```diff
@@ -0,0 +1,160 @@
+	private function maybe_redirect_to_edit_order_page(): void {
+		$post_id = absint( $_GET['post'] ?? 0 );
+
+		if ( ! $post_id || 'shop_order' !== get_post_type( $post_id ) || ! isset( $_GET['action'] ) ) {
```
Should we also test for posts of type `shop_order_placehold`?

(I'm thinking here of cases where some kind of third party app generates these URLs based on its knowledge of order IDs, rather than cases where a human user captured and preserved a historic link.)

jorgeatorres replied:
Yes. That definitely makes sense.

barryhughes reviewed (commented):
Thinking of edit URLs and bulk action URLs: we don't set up the redirect hooks unless [custom orders table usage is enabled](https://github.com/woocommerce/woocommerce/pull/34644/files#diff-21638fa99f423148e7c894760eb4990fa94f6b1ed77180c6522937044b2dafd0R27-R29). However, what if usage is enabled but the migration is incomplete (using the example of edit URLs, what if the post in question has not yet been migrated)?

jorgeatorres commented:
Hi @barryhughes!

Thanks for the review.

> The only thing I am having trouble with are bulk actions (such as trashing orders).

I didn't really cover this scenario. I was thinking it wasn't really that common for 3rd parties to generate bulk action URLs (or for users to save temporarily save those) but we do support individual trashing so it wouldn't hurt to support bulk actions in general.
I'll make the relevant changes.

jorgeatorres commented:
> Thinking of edit URLs and bulk action URLs: we don't set up the redirect hooks unless [custom orders table usage is enabled](https://github.com/woocommerce/woocommerce/pull/34644/files#diff-21638fa99f423148e7c894760eb4990fa94f6b1ed77180c6522937044b2dafd0R27-R29). However, what if usage is enabled but the migration is incomplete (using the example of edit URLs, what if the post in question has not yet been migrated)?

My understanding was that you couldn't switch the authoritative datastore until after all orders had been migrated, but I might be missing something.
Can you think of a particular scenario where one could end up with a post with no corresponding order in the COT tables and with COT enabled?

jorgeatorres committed:
Update changelog

jorgeatorres committed:
Take into account ‘shop_orderplacehold’ when redirecting

jorgeatorres committed:
Correctly handle bulk actions

jorgeatorres commented:
Hi @barryhughes!

I've addressed the feedback, so please review again when you have the chance.
I'm pretty sure you are already doing something similar, but just in case it's helpful, here's how I'm testing:
- Make COT authoritative.
- Comment out [this line](https://github.com/woocommerce/woocommerce/blob/5e57471c5aa7ebf72f2a41153d3a586c75f63c0a/plugins/woocommerce/src/Internal/Admin/Orders/PageController.php#L132) to see both the CPT and COT "Orders" link under the WC menu.
- Temporarily comment out [this other line](https://github.com/woocommerce/woocommerce/blob/8003a88e0e7d2adddef2f254695a1f1adc87311c/plugins/woocommerce/src/Internal/Admin/Orders/PostsRedirectionController.php#L34) so that I don't get automatically redirected to the COT table and then toggle the commented line before clicking a link or performing a bulk action.

Anyways, please let me know what you think. Thanks!

barryhughes reviewed (approved):
Nice!

barryhughes commented:
I believe the E2E fails are unrelated, but giving them an extra spin before merging (PHP 7.4/nightly fails are almost certainly unrelated, they seem to be impacting most PRs).

jorgeatorres commented:
> I believe the E2E fails are unrelated, but giving them an extra spin before merging (PHP 7.4/nightly fails are almost certainly unrelated, they seem to be impacting most PRs).

Yep. That's being discussed: p1662969426268289-slack-C03CPM3UXDJ.
